module/localisation/BallLocalisation/src/BallLocalisation.cpp

```diff
@@ -1,40 +1,120 @@
+            last_time_update_time = curr_time;
+            filter.timeUpdate(seconds);
+
+            /* Creating ball state vector and covariance matrix for emission */
+            //std::unique_ptr ball;
+            auto ball = std::make_unique<Ball>();
+            ball->locObject.position = convert<double,2>(filter.get());
+            ball->locObject.position_cov = convert<double,2,2>(filter.getCovariance());
+            ball->locObject.last_measurement_time = last_measurement_update_time;
+
+            /* For graphing purposes */
+            arma::vec3 rBWw = { ball->locObject.position[0], ball->locObject.position[1], field.ball_radius };
+            // Get our transform to world coordinates
+            const Transform3D& Htw = convert<double, 4, 4>(sensors.world);
+            const Transform3D& Htc = convert<double, 4, 4>(sensors.forwardKinematics.at(ServoID::HEAD_PITCH));
+            Transform3D Hcw = Htc.i() * Htw;
+            arma::vec3 rBCc_cart = Hcw.transformPoint(rBWw);
+            arma::vec3 rBCc_sph1 = cartesianToSpherical(rBCc_cart); // in r,theta,phi
+            arma::vec3 rBCc_sph2 = { rBCc_sph1[0], rBCc_sph1[1], rBCc_sph1[2] };  // in roe, theta, phi, where roe is 1/r
+            emit(graph("ball loc state", rBCc_sph2[0], rBCc_sph2[1],rBCc_sph2[2]));
+
+            emit(graph("ball state xy",ball->locObject.position[0],ball->locObject.position[1]));
+            emit(ball);
+        });
+
+        /* To run whenever a ball has been detected */
+        on<Trigger<std::vector<message::vision::Ball>>
+         , With<FieldDescription>
+         , With<Sensors>>().then([this](
+            const std::vector<message::vision::Ball>& balls
+            , const FieldDescription& field
+            , const Sensors& sensors){
+
+                double quality = 1.0;   // I don't know what quality should be used for
+                if(balls.size() > 0){
+
+                    /* Call Time Update first */
+                    auto curr_time = NUClear::clock::now();
+                    //TODO: take min of time and measurement update delta
+                    double seconds = TimeDifferenceSeconds(curr_time,last_time_update_time);
+                    last_time_update_time = curr_time;
+                    filter.timeUpdate(seconds);
+
+                    /* Now call Measurement Update. Supports multiple measurement methods and will treat them as separate measurements */
+                    for (auto& measurement : balls[0].measurements) {
+                        /* For graphing purposes */
+                        arma::vec3 rBCc = convert<double, 3, 1>(measurement.rBCc);
+                        //TODO: make this be in ro coords to start with
+                        rBCc[0] = rBCc[0];
+                        emit(graph("ball meas", rBCc[0],rBCc[1], rBCc[2]));
+
+                        quality *= filter.measurementUpdate(rBCc,convert<double, 3, 3>(measurement.covariance), field, sensors);
```

Comment:
You shouldn't need this quality. It's normally used for multi modal stuff, it'll tell you how likely the measurement was. You could use it if you had multiple vision balls to work out which is the most likely correct.